test/e2e/federate.go

```diff
@@ -0,0 +1,184 @@
+				// Check that placement is set to all available clusters
+				validatePlacement(tl, fedResource, clusterNames)
+			})
+		})
+	}
+})
+
+func createTargetResource(tl common.TestLogger, kubeConfig *restclient.Config, typeConfig typeconfig.Interface, typeConfigName, namespace string, fixture *unstructured.Unstructured) *unstructured.Unstructured {
+	targetAPIResource := typeConfig.GetTarget()
+	targetTypeClient, err := util.NewResourceClient(kubeConfig, &targetAPIResource)
+	if err != nil {
+		tl.Fatalf("Error creating target resource client for %q: %v", typeConfigName, err)
+	}
+	targetResource, err := getTargetResource(targetAPIResource, fixture, typeConfigName, namespace)
+	if err != nil {
+		tl.Fatalf("Error getting target resource for %q: %v", typeConfigName, err)
+	}
+	createdResource, err := targetTypeClient.Resources(namespace).Create(targetResource, metav1.CreateOptions{})
+	if err != nil {
+		tl.Fatalf("Error creating target resource for %q: %v", typeConfigName, err)
+	}
+
+	return createdResource
+}
+
+func getTargetResource(targetAPIResource metav1.APIResource, fixture *unstructured.Unstructured, typeConfigName, namespace string) (*unstructured.Unstructured, error) {
+	targetResource := &unstructured.Unstructured{}
+	template, ok, err := unstructured.NestedFieldCopy(fixture.Object, ctlutil.TemplateField)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Error retrieving template field for %q fixture: %v", typeConfigName, err)
+	}
+	if ok {
+		targetResource.Object = template.(map[string]interface{})
+		federate.SetBasicMetaFields(targetResource, targetAPIResource, "", namespace, "test-e2e-")
+		return targetResource, nil
+	}
+
+	return nil, errors.New("Error: template from fixture is empty")
+}
+
+func waitForFederatedResource(tl common.TestLogger, kubeConfig *restclient.Config, typeConfig typeconfig.Interface, desiredTemplate *unstructured.Unstructured, namespace string) *unstructured.Unstructured {
+	var federatedResource *unstructured.Unstructured
+	fedAPIResource := typeConfig.GetFederatedType()
+	fedClient, err := util.NewResourceClient(kubeConfig, &fedAPIResource)
+	err = wait.PollImmediate(framework.PollInterval, framework.TestContext.SingleCallTimeout, func() (exist bool, err error) {
+		federatedResource, err = fedClient.Resources(namespace).Get(desiredTemplate.GetName(), metav1.GetOptions{})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+
+		return validatEquivalence(tl, federatedResource, desiredTemplate), nil
+	})
+
+	if err != nil {
+		tl.Fatalf("Error checking equivalance: %v", err)
+	}
+	return federatedResource
+}
+
+func validatEquivalence(tl common.TestLogger, fedResource, desiredTemplate *unstructured.Unstructured) bool {
+	a, ok, err := unstructured.NestedFieldCopy(fedResource.Object, ctlutil.SpecField, ctlutil.TemplateField)
+	if err != nil || !ok {
+		tl.Fatal("Error retriving template from actual federated resource")
```

Comment:
s/retriving/retrieving/

Reply:
The federated test is included with `crud` now!